### DISCUSSION TITLE
Cannot build second-order dependent packages?

### DISCUSSION THREAD

**Discussion by prescience-data** [supabase/supabase#3368] [Questions] [answered]
We have a central package that defines all of our Supabase actions _(such as `updateAccount()` for example)_ which provides these functions to the rest of the packages in our org.

When attempting to build a package that depends on this intermediate package (eg  `@my-org/supabase`), the following errors for `@supabase/gotrue-js`, `@supabase/storage-js`, `@supabase/postgrest-js` are thrown.

Building the core Supabase `@my-org/supabase` function package works fine, the issue is only when building downstream packages that have `@my-org/supabase` as a dependency. 

Importantly, this issue does not occur with any other dependencies other than Supabase (ie we have many other similar packages like `@my-org/logging` etc that provide core functions in a similar way).

### Stuff we've tried:

- Adding all Supabase packages to the `"types": []` section in `tsconfig.json`.
- Copying all the individual declaration files by hand into the central package and exporting them.
- Adding `/// <reference types="..." />` lines to the central package export.

I feel like I'm approaching the edges of my TypeScript troubleshooting knowledge here, and I'm sure someone other than me has attempted to do this so would massively appreciate any threads to pull on in terms of possible solutions. 🙏🏻

My final possible strategy is to bundle the package with `esbuild`, however this would create a huge amount of redundant code in our final bundles, as many of the packages required by the core Supabase package have overlap with upstream modules. 
This causes lot of overlapping code to be duplicated, which sucks given that Container size is a big concern for us, so tree-shaking is crucial for optimization. 😔

### Error Dump

```
[typescript] Error: ../../common/temp/node_modules/.pnpm/@supabase+gotrue-js@1.18.0/node_modules/@supabase/gotrue-js/dist/main/GoTrueClient.d.ts:8:55 - (TS2304) Cannot find name 'Storage'.
[typescript] Error: ../../common/temp/node_modules/.pnpm/@supabase+postgrest-js@0.33.3/node_modules/@supabase/postgrest-js/dist/main/lib/types.d.ts:43:20 - (TS2304) Cannot find name 'URL'.
[typescript] Error: ../../common/temp/node_modules/.pnpm/@supabase+storage-js@1.4.0/node_modules/@supabase/storage-js/dist/main/lib/StorageFileApi.d.ts:35:84 - (TS2304) Cannot find name 'File'.
[typescript] Error: ../../common/temp/node_modules/.pnpm/@supabase+storage-js@1.4.0/node_modules/@supabase/storage-js/dist/main/lib/StorageFileApi.d.ts:35:91 - (TS2304) Cannot find name 'FormData'.
[typescript] Error: ../../common/temp/node_modules/.pnpm/@supabase+storage-js@1.4.0/node_modules/@supabase/storage-js/dist/main/lib/StorageFileApi.d.ts:35:126 - (TS2304) Cannot find name 'ReadableStream'.
[typescript] Error: ../../common/temp/node_modules/.pnpm/@supabase+storage-js@1.4.0/node_modules/@supabase/storage-js/dist/main/lib/StorageFileApi.d.ts:35:155 - (TS2304) Cannot find name 'URLSearchParams'.
[typescript] Error: ../../common/temp/node_modules/.pnpm/@supabase+storage-js@1.4.0/node_modules/@supabase/storage-js/dist/main/lib/StorageFileApi.d.ts:51:84 - (TS2304) Cannot find name 'File'.
[typescript] Error: ../../common/temp/node_modules/.pnpm/@supabase+storage-js@1.4.0/node_modules/@supabase/storage-js/dist/main/lib/StorageFileApi.d.ts:51:91 - (TS2304) Cannot find name 'FormData'.
[typescript] Error: ../../common/temp/node_modules/.pnpm/@supabase+storage-js@1.4.0/node_modules/@supabase/storage-js/dist/main/lib/StorageFileApi.d.ts:51:126 - (TS2304) Cannot find name 'ReadableStream'.
[typescript] Error: ../../common/temp/node_modules/.pnpm/@supabase+storage-js@1.4.0/node_modules/@supabase/storage-js/dist/main/lib/StorageFileApi.d.ts:51:155 - (TS2304) Cannot find name 'URLSearchParams'.
  ...2 lines omitted...
  [typescript] ../../common/temp/node_modules/.pnpm/@supabase+postgrest-js@0.33.3/node_modules/@supabase/postgrest-js/dist/main/lib/types.d.ts:43:20 - (TS2304) Cannot find name 'URL'.
  [typescript] ../../common/temp/node_modules/.pnpm/@supabase+storage-js@1.4.0/node_modules/@supabase/storage-js/dist/main/lib/StorageFileApi.d.ts:35:84 - (TS2304) Cannot find name 'File'.
  [typescript] ../../common/temp/node_modules/.pnpm/@supabase+storage-js@1.4.0/node_modules/@supabase/storage-js/dist/main/lib/StorageFileApi.d.ts:35:91 - (TS2304) Cannot find name 'FormData'.
  [typescript] ../../common/temp/node_modules/.pnpm/@supabase+storage-js@1.4.0/node_modules/@supabase/storage-js/dist/main/lib/StorageFileApi.d.ts:35:126 - (TS2304) Cannot find name 'ReadableStream'.
  [typescript] ../../common/temp/node_modules/.pnpm/@supabase+storage-js@1.4.0/node_modules/@supabase/storage-js/dist/main/lib/StorageFileApi.d.ts:35:155 - (TS2304) Cannot find name 'URLSearchParams'.
  [typescript] ../../common/temp/node_modules/.pnpm/@supabase+storage-js@1.4.0/node_modules/@supabase/storage-js/dist/main/lib/StorageFileApi.d.ts:51:84 - (TS2304) Cannot find name 'File'.
  [typescript] ../../common/temp/node_modules/.pnpm/@supabase+storage-js@1.4.0/node_modules/@supabase/storage-js/dist/main/lib/StorageFileApi.d.ts:51:91 - (TS2304) Cannot find name 'FormData'.
  [typescript] ../../common/temp/node_modules/.pnpm/@supabase+storage-js@1.4.0/node_modules/@supabase/storage-js/dist/main/lib/StorageFileApi.d.ts:51:126 - (TS2304) Cannot find name 'ReadableStream'.
  [typescript] ../../common/temp/node_modules/.pnpm/@supabase+storage-js@1.4.0/node_modules/@supabase/storage-js/dist/main/lib/StorageFileApi.d.ts:51:155 - (TS2304) Cannot find name 'URLSearchParams'.
Error: Encountered TypeScript errors
```

**Comment by prescience-data** [comment, marked as answer]
Ok so it seems to be because I am not including DOM in the second-order projects' `tsconfig` libs.
Given this is a fully server-side Node project, I'd prefer not to have to implement those extra libs if possible - but happy to capitulate if there is no other way.

**Reply by soedirgo** [reply]
Just to clarify - are these are all type errors rather than logic errors?

There are some types that are used to allow browser-only things as well as Node.js-only things, e.g. a "file" for storage-js can be a `ReadableStream` (browser-only) or a `NodeJS.ReadableStream` (Node.js-only). They're argument types though, so you don't need to implement polyfills for what you don't use. `lib` in `tsconfig` afaik is only for type definitions, not implementations.

**Reply by prescience-data** [reply]
It's a type issue that blocks building. It can be resolved by adding the `DOM` types to the `lib` section of `tsconfig` which is not a huge deal (as you say, it's only for defs) - it was just confusing the team as they are Node packages in this case so hadn't thought to add those DOM types to the next order packages. Closing for now as this is a viable fix - thanks for having a look :D

**Reply by soedirgo** [reply]
Thanks, sorry for the inconvenience 🙏
